### PR TITLE
Allow specifying S3 KMS Key ID via security mapping config and extra credentials

### PR DESCRIPTION
**Pull request by losipiuk**

### CODE REVIEW

**Comment by losipiuk** [comment]
Replacement for https://github.com/trinodb/trino/pull/6615

**Comment by losipiuk** [comment]
@findepi @electrum PTAL

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMappingConfigurationProvider.java [hunk 1]
```diff
@@ -109,6 +110,11 @@ public void updateConfiguration(Configuration configuration, HdfsContext context
             hasher.putString(role, UTF_8);
         });
 
+        mapping.getKmsKeyId().ifPresent(key -> {
+            configuration.set(S3_KMS_KEY_ID, key);
+            hasher.putString(key, UTF_8);
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMappingConfigurationProvider.java:118]: add "kms key id" to hash as well (as per previous GK's comment)

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMapping.java [hunk 2]
```diff
@@ -74,6 +76,9 @@ public S3SecurityMapping(
 
         this.kmsKeyId = requireNonNull(kmsKeyId, "kmsKeyId is null");
 
+        this.allowedKmsKeyIds = requireNonNull(allowedKmsKeyIds, "allowedKmsKeyIds is null")
+                .orElse(ImmutableList.of());
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMapping.java:80]: make defensive copy when list is present

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMappingConfigurationProvider.java [hunk 3]
```diff
@@ -155,4 +158,29 @@ public void updateConfiguration(Configuration configuration, HdfsContext context
         }
         return extraCredentialRole;
     }
+
+    private Optional<String> selectKmsKeyId(S3SecurityMapping mapping, HdfsContext context)
+    {
+        Optional<String> optionalSelected = getKmsKeyIdFromExtraCredential(context);
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMappingConfigurationProvider.java:164]: optionalSelected -> userProvided, userSelected

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMappingConfigurationProvider.java [hunk 4]
```diff
@@ -155,4 +158,29 @@ public void updateConfiguration(Configuration configuration, HdfsContext context
...
+        if (optionalSelected.isEmpty()) {
+            return mapping.getKmsKeyId();
+        }
+
+        String selected = optionalSelected.get();
+
+        // selected key id must match default or be allowed
+        if (!selected.equals(mapping.getKmsKeyId().orElse(null)) &&
+                !mapping.getAllowedKmsKeyIds().contains(selected) &&
+                !mapping.getAllowedKmsKeyIds().contains(ANY_KMS_KEY_ID)) {
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMappingConfigurationProvider.java:175]: Make `getAllowedKmsKeyIds` List ->  Set

**Review by findepi — APPROVED** [review]

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMappingConfigurationProvider.java (hunk 1 above)
Comment (losipiuk) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMappingConfigurationProvider.java:118]: Oh - good catch - forgot about that. Sorry

Comment (losipiuk) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMappingConfigurationProvider.java:118]: Technically the comment also applies for other entries put in hasher. Though crafting a collision should not be something observable in the real world.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMappingConfigurationProvider.java (hunk 4 above)
Comment (losipiuk) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMappingConfigurationProvider.java:175]: done

**Comment by losipiuk** [comment]
@electrum do you want to take a look?

**Pull request merged**